torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+              make sure it is idempotent in that case.
+        """
+        for p in self.params:
+            assert p._full_param_padded.storage().size() != 0  # type: ignore[attr-defined]
+            p.data = p._full_param_padded[: p._orig_size.numel()].view(p._orig_size)  # type: ignore[attr-defined]
+
+    @torch.no_grad()
+    def _prep_grads_for_backward(self) -> None:
+        """Make sure p.grad has the correct size/device, otherwise set it to None."""
+        for p in self.params:
+            if p.grad is not None and (
+                p.grad.size() != p._orig_size  # type: ignore[attr-defined]
+                or p.grad.device != p.device
+            ):
+                p.grad = None
+
+    @torch.no_grad()
+    def _free_full_params(self, params: Optional[List[Parameter]] = None) -> None:
+        """Free up storage for full parameters."""
+        if params is None:
+            params = self.params
+        current_stream = torch.cuda.current_stream()
+        for p in params:
+            # e.g., world_size == 1
+            if not p._is_sharded:  # type: ignore[attr-defined]
+                continue
+            # Don't let PyTorch reuse this memory until all work in the current
+            # stream is complete.
+            p._full_param_padded.record_stream(current_stream)  # type: ignore[attr-defined]
+            # There may be external references to the Tensor Storage that we
+            # can't modify, such as references that are created by
+            # ctx.save_for_backward in the forward pass. Thus when we
+            # unshard parameters, we should reuse the original Tensor
+            # Storage object and unshard it in-place. For now, just resize
+            # the Storage to 0 to save memory.
+            _free_storage(p._full_param_padded)  # type: ignore[attr-defined]
+
+
+    @torch.no_grad()
+    def _use_param_local_shard(self, params: Optional[List[Parameter]] = None) -> None:
+        """Use local shard for a list of params."""
+        if params is None:
+            params = self.params
+        for p in params:
+            p.data = p._local_shard  # type: ignore[attr-defined]
+
+    def _assert_state(self, state: Union[TrainingState, List[TrainingState]]) -> None:
+        """Assert we are in the given state."""
+        # Since assert can be turned off and this error checking
+        # is really important, we use explicit error checking
+        # and raise a ValueError if needed.
+        if isinstance(state, TrainingState):
+            state = [state]
+        if self.training_state not in state:
+            msg = f"expected to be in states {state} but current state " f"is {self.training_state}"
+            # In case we are failing in the context of autograd hook, asserting
+            # may not generate useful msg. So, let's print it to be sure.
+            if self.rank == 0:
+                print(f"Asserting FSDP instance is: {self}")
```

Comment:
Why the printing here? The error below should already do that?

Reply:
The "state does not match" error is threw in each rank. Printing more verbosed debugging message on rank 0 only.